File: snacking_problem-20/solution.cpp
#include <iostream>
#include <map>
#include <string>

using namespace std;

int main() {
  map<int, string> valueMap;
  map<string, int> keyMap;
  int total, current, num, lookingFor;
  string name;
  cin >> total;
  cin >> num;
  for(int i = 0; i < num; i++){
    cin >> name >> current;
    valueMap[current] = name;
    keyMap[name] = current;
  }
  cin >> num;
  for(int i = 0; i < num ; i++){
    cin >> name;
    current = keyMap[name];
    lookingFor = total - current;
    auto it = valueMap.find(lookingFor);
    if(it != valueMap.end()){
      cout << valueMap[lookingFor] << endl;
    }else{
      cout << "There is no matching snack Joe" << endl;
    }
  }
  return 0;
}

File: snacking_problem-20/solution.py
#!/usr/bin/python
import sys

total = raw_input()
readin = raw_input()

data = dict()
kdata = dict()

for x in xrange(0, int(readin)):
    tmp = raw_input().split(" ")
    data[int(tmp[1])] = tmp[0]
    kdata[str(tmp[0])] = int(tmp[1])

readin = raw_input()

for x in xrange(0, int(readin)):
    tmp = raw_input()
    pval = int(total) - kdata[tmp]
    if pval in data.keys():
        sys.stdout.write(data[pval])
        if x < int(readin) - 1:
            sys.stdout.write("\n")
    else:
        sys.stdout.write("There is no matching snack Joe")
        if x < int(readin) - 1:
            sys.stdout.write("\n")